lib/services/access_list.go

```diff
@@ -243,20 +258,147 @@ func IsAccessListOwner(identity tlsca.Identity, accessList *accesslist.AccessLis
 // AccessListMembershipChecker will check if users are members of an access list and
 // makes sure the user is not locked and meets membership requirements.
 type AccessListMembershipChecker struct {
-	members AccessListMemberGetter
+	members AccessListsAndMembersGetter
 	locks   LockGetter
 	clock   clockwork.Clock
 }
 
 // NewAccessListMembershipChecker will create a new access list membership checker.
-func NewAccessListMembershipChecker(clock clockwork.Clock, members AccessListMemberGetter, locks LockGetter) *AccessListMembershipChecker {
+func NewAccessListMembershipChecker(clock clockwork.Clock, members AccessListsAndMembersGetter, locks LockGetter) *AccessListMembershipChecker {
 	return &AccessListMembershipChecker{
 		members: members,
 		locks:   locks,
 		clock:   clock,
 	}
 }
 
+func getAccessListDynamicMembers(ctx context.Context, clt AccessListsAndMembersGetter, entry string) ([]string, error) {
+	var pageToken string
+	var dynamicMembers []string
+
+	for {
+		var members []*accesslist.AccessListMember
+		var err error
+
+		members, pageToken, err = clt.ListAccessListMembers(ctx, entry, 0 /* default page size */, pageToken)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		for _, member := range members {
+			if member.Spec.Kind == accesslist.MemberKindList {
+				dynamicMembers = append(dynamicMembers, member.GetName())
+			}
+		}
+
+		if pageToken == "" {
+			break
+		}
+	}
+
+	return dynamicMembers, nil
+}
+
+func (a AccessListMembershipChecker) recursiveIsAccessListMemberCheck(ctx context.Context, identity tlsca.Identity, accessList *accesslist.AccessList) error {
+	seen := map[string]struct{}{}
+	queue := []string{accessList.GetName()}
+
+	for len(queue) > 0 {
+		size := len(queue)
+		for i := 0; i < size; i++ {
+			pal := queue[0]
+			queue = queue[1:]
```

Comment:
Can we simplify this to: ? 
```suggestion
	for len(queue) > 0 {
		pal := queue[0]
		queue = queue[1:]
```